pubsub/azurepubsub/azurepubsub.go

```diff
@@ -0,0 +1,438 @@
+	ns, err := getSbNamespace(connectionString)
+	if err != nil {
+		return err
+	}
+
+	subManager, err := ns.NewSubscriptionManager(topicName)
+	if err != nil {
+		return err
+	}
+	
+	_, err = ensureSubscription(ctx, subManager, subscriptionName, opts...)	
+	return err	
+}
+
+func ensureSubscription(ctx context.Context, sm *servicebus.SubscriptionManager, name string, opts ...servicebus.SubscriptionManagementOption) (*servicebus.SubscriptionEntity, error) {
+	subEntity, err := sm.Get(ctx, name)
+	if err == nil {
+		_ = sm.Delete(ctx, name)
+	}
+
+	subEntity, err = sm.Put(ctx, name, opts...)
+	if err != nil {		
+		return nil, err
+	}
+
+	return subEntity, nil
+}
+
+func deleteSubscription(ctx context.Context, topicName string, subscriptionName string, connectionString string) (error){
+	ns, err := getSbNamespace(connectionString)
+	if err != nil {
+		return err
+	}
+	sm, err := ns.NewSubscriptionManager(topicName)
+	se, err := sm.Get(ctx, subscriptionName)
+	if se != nil {
+		_ = sm.Delete(ctx, subscriptionName)
+	}
+	return nil
+}
+
+func (s *subscription) getSbSubscription(ctx context.Context)(*servicebus.Subscription, error){
+	sm, err := s.ns.NewSubscriptionManager(s.topicName)	
+	if err != nil {
+		return nil, err
+	}
+	// An empty SubscriptionEntity means no subscription exist for the given name. 
+	se, err := sm.Get(ctx, s.name)
+	if se == nil {
+		return nil, fmt.Errorf("azurepubsub: no such subscription %v", s.name)
+	}
+
+	sbTopic, err := s.ns.NewTopic(s.topicName)	
+	if err != nil {
+		return nil, err
+	}
+	return sbTopic.NewSubscription(s.name)
+}
+
+// IsRetryable implements driver.Subscription.IsRetryable.
+func (s *subscription) IsRetryable(error) bool {	
+	return false
+}
+
+// As implements driver.Subscription.As.
+func (s *subscription) As(i interface{}) bool {
+	p, ok := i.(**servicebus.Subscription)
+	if !ok {
+		return false
+	}	
+	*p, _ = s.getSbSubscription(context.Background())
+	return true
+}
+
+
+func (s *subscription) ErrorCode(err error) gcerrors.ErrorCode {
+	return errorCode(err)
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {			
+	sbSub, err := s.getSbSubscription(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func(){
+		sbSub.Close(ctx)
+	}()
+	
+	receiverCtx, cancelCtx := context.WithTimeout(ctx, s.opts.ListenerTimeout)
+	defer cancelCtx()
+	
+	subReceiver, err := sbSub.NewReceiver(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func(){
+		subReceiver.Close(ctx)
+	}()
+	
+	var messages []*driver.Message			
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		subReceiver.Listen(receiverCtx, servicebus.HandlerFunc(func(iCtx context.Context, sbmsg *servicebus.Message)(error){		
+			metadata := map[string]string{}
+			for k, v := range sbmsg.UserProperties {			
+				metadata[k] = v.(string)
```

Comment:
I'll let the SB SDK handle it with the [ForEachKey ](https://github.com/Azure/azure-service-bus-go/blob/7dafc265e8a87e9d1b4070e6f4a8dad19416977a/message.go#L310)function. Its treated as string but can be an object, for example it was set by the .NET SDK. Normally, the user should own serialization/deserialization of complex types used in Custom Properties. BTW, perhaps we are missing a Message.As?; If so, this will solve a couple of my issues including Ack & Custom Types in Metadata.

Reply:
We're debating what Message.As should look like in #657. We'd love your thoughts.